### PR TITLE
Not support anonymous union in struct #317

### PR DESCRIPTION
**Pull request by alexey-utkin**

### CODE REVIEW

**Review by alexey-utkin — COMMENTED** [review]
Fixing
[PERF] UTBot consumes CPU in spinlocks #382

File: server/src/Server.cpp [hunk 1]
```diff
@@ -420,8 +420,9 @@ Status Server::TestsGenServiceImpl::provideLoggingCallbacks(
          * 1. Using gRPC async API
          * 2. Issuing a request from UTBot to a specific client on every log entry.
          */
+        using namespace std::chrono_literals;
         while (holdLockFlag[callbackName].exchange(true, std::memory_order_acquire)) {
-            std::this_thread::yield();
+            std::this_thread::sleep_for(100ms);
```
Comment (alexey-utkin) [inline server/src/Server.cpp:425]: [PERF] UTBot consumes CPU in spinlocks #382

File: server/src/Tests.h [hunk 2]
```diff
@@ -674,33 +662,29 @@ namespace tests {
...
+        static std::shared_ptr<EnumValueView> enumView(const std::vector<char> &byteArray,
+                                                       const types::EnumInfo &enumInfo,
+                                                       size_t offset,
+                                                       size_t len);
+
         std::shared_ptr<PrimitiveValueView> primitiveView(const std::vector<char> &byteArray,
                                                           const types::Type &type,
                                                           size_t offset,
                                                           size_t len);
-        static std::shared_ptr<EnumValueView> enumView(const std::vector<char> &byteArray,
```
Comment (alexey-utkin) [inline server/src/Tests.h:700]: Group enum and struct-like. Seems they need to be joined in the future.

File: server/src/printers/KleeConstraintsPrinter.cpp [hunk 3]
```diff
@@ -50,7 +47,7 @@ void KleeConstraintsPrinter::genConstraintsForPrimitive(const ConstraintsState &
     if (!cons.empty()) {
         strFunctionCall(PrinterUtils::KLEE_PREFER_CEX, { state.paramName, cons });
     } else {
-        ss << TAB_N() << "// No constraints for " << state.curElement << NL;
+        ss << LINE_INDENT() << "// No constraints for " << state.curElement << NL;
```
Comment (alexey-utkin) [inline server/src/printers/KleeConstraintsPrinter.cpp:50]: Indent can include the inline comment, so rename it to indent.

File: server/src/printers/KleeConstraintsPrinter.cpp [hunk 4]
```diff
@@ -148,37 +111,42 @@ void KleeConstraintsPrinter::genConstraintsForPointerInStruct(const ConstraintsS
...
+    bool isStruct = curStruct.subType == SubType::Struct;
     for (const auto &field : curStruct.fields) {
         std::string errorMessage = "Unrecognized field of type '" + field.type.typeName() +
                                    "' in struct '" + curStruct.name + "'.";
         auto access = PrinterUtils::getFieldAccess(state.curElement, field);
-        ConstraintsState newState = { state.paramName, access, field.type, state.endString, state.depth + 1 };
+        ConstraintsState newState = { state.paramName,
+                                      access,
+                                      field.type,
+                                      isStruct ? state.endString : false,
```
Comment (alexey-utkin) [inline server/src/printers/KleeConstraintsPrinter.cpp:122]: That is still essential difference between struct and union. Need to be investigated (in progress).

File: server/src/printers/KleeConstraintsPrinter.cpp [hunk 5]
```diff
@@ -148,37 +111,42 @@ void KleeConstraintsPrinter::genConstraintsForPointerInStruct(const ConstraintsS
...
+        case TypeKind::PRIMITIVE:
+            genConstraintsForPrimitive(newState);
+            break;
+        case TypeKind::ARRAY:
+            genConstraintsForPointerOrArray(newState);
             break;
         case TypeKind::OBJECT_POINTER:
             if (types::TypesHandler::isArrayOfPointersToFunction(field.type)) {
-                genStubForStructFunctionPointerArray(state.curElement, field.name, stubFunctionName);
+                genStubForStructFunctionPointerArray(state.curElement, field,
```
Comment (alexey-utkin) [inline server/src/printers/KleeConstraintsPrinter.cpp:142]: no more `just by name` access. That leads to unpredictable errors.

File: server/src/printers/TestsPrinter.cpp [hunk 6]
```diff
@@ -705,17 +705,33 @@ std::string printer::MultiLinePrinter::print(TestsPrinter *printer,
     std::stringstream structuredValuesWithPrefixes;
     structuredValuesWithPrefixes << "{" << NL;
     ++printer->tabsDepth;
+
+    int longestFieldIndexForUnionInit = view->getLongestFieldIndexForUnionInit();
```
Comment (alexey-utkin) [inline server/src/printers/TestsPrinter.cpp:709]: So now we are printing all the fields in union, but commenting all, except the longest.

File: server/src/types/Types.cpp [hunk 7]
```diff
@@ -864,12 +834,6 @@ types::TypesHandler::isSupportedType(const Type &type, TypeUsage usage, int dept
               return getTypeKind(type) == TypeKind::UNKNOWN;
             }
         },
-        {
```
Comment (alexey-utkin) [inline server/src/types/Types.cpp:867]: the subject of bugfix.

File: server/src/types/Types.cpp [hunk 8]
```diff
@@ -894,14 +858,10 @@ types::TypesHandler::isSupportedType(const Type &type, TypeUsage usage, int dept
                 });
               };
 
-              if (isUnion(type)) {
```
Comment (alexey-utkin) [inline server/src/types/Types.cpp:897]: Has no direct influence to the tests

File: server/src/types/Types.cpp [hunk 9]
```diff
@@ -954,15 +914,10 @@ types::TypesHandler::isSupportedType(const Type &type, TypeUsage usage, int dept
                   return false;
                 });
               };
-              if (isStruct(type)) {
```
Comment (alexey-utkin) [inline server/src/types/Types.cpp:957]: Has no direct influence to the tests

File: server/src/types/Types.cpp [hunk 10]
```diff
@@ -1017,8 +972,8 @@ types::Type types::TypesHandler::getReturnTypeToCheck(const types::Type &returnT
     return returnType;
 }
 
-std::string types::EnumInfo::getEntryName(const std::string &value, utbot::Language language) {
-    auto const& entry = valuesToEntries[value];
+std::string types::EnumInfo::getEntryName(const std::string &value, utbot::Language language) const {
+    const EnumEntry &entry = valuesToEntries.at(value);
```
Comment (alexey-utkin) [inline server/src/types/Types.cpp:976]: Has no const `operator[]`, so changed.

**Review by ladisgin — CHANGES_REQUESTED** [review]

File: server/src/Tests.cpp [hunk 11]
```diff
@@ -1129,13 +1037,18 @@ Tests::MethodDescriptionHash::operator()(const Tests::MethodDescription &methodD
...
 }
 
-UnionValueView::UnionValueView(
-    const std::string &typeName,
-    const std::shared_ptr<AbstractValueView> &rawDataView,
-    std::vector<std::shared_ptr<AbstractValueView>, std::allocator<std::shared_ptr<AbstractValueView>>> subViews)
-    : AbstractValueView(std::move(subViews)),
-      entryValue(PrinterUtils::convertBytesToUnion(typeName, rawDataView->getEntryValue(nullptr))) {
-}
+//UnionValueView::UnionValueView(
```
[the remark this answers is not in the document]
Comment (alexey-utkin) [reply server/src/Tests.cpp:1040]: done

File: server/src/types/Types.h [hunk 12]
```diff
@@ -523,20 +528,11 @@ namespace types {
...
         EnumInfo getEnumInfo(const Type&) const;
 
-        /**
-         * Returns UnionInfo by given union name.
-         * For safe usage, please use isUnion(..) before calling getUnionInfo(..).
-         * @return UnionInfo for given union.
-         */
-        UnionInfo getUnionInfo(const Type&) const;
-
         bool isStruct(uint64_t id) const;
```
[the remark this answers is not in the document]
Comment (alexey-utkin) [reply server/src/types/Types.h:531]: done

File: server/src/types/Types.h [hunk 13]
```diff
@@ -280,17 +280,18 @@ namespace types {
...
 
+    enum class SubType {
+        Struct,
+        Union
+    };
+
     struct StructInfo: TypeInfo {
         std::vector<Field> fields{};
-
+        int longestFieldIndexForUnionInit;
```
[the remark this answers is not in the document]
Comment (alexey-utkin) [reply server/src/types/Types.h]: done

File: server/src/printers/TestsPrinter.cpp [hunk 14]
```diff
@@ -705,17 +705,33 @@ std::string printer::MultiLinePrinter::print(TestsPrinter *printer,
...
+            if (longestFieldIndexForUnionInit < 0)
+                structuredValuesWithPrefixes << "," << NL;
+            else
+                structuredValuesWithPrefixes << NL;
+        }
+        if (longestFieldIndexForUnionInit < 0 || longestFieldIndexForUnionInit == i) {
+            structuredValuesWithPrefixes << printer->LINE_INDENT() << view->getFieldPrefix(i)
+                                         << sview->getEntryValue(printer);
+        }
+        else {
```
[the remark this answers is not in the document]
Comment (alexey-utkin) [reply server/src/printers/TestsPrinter.cpp]: done

**Review by ladisgin — APPROVED** [review]

**Pull request merged**